### ISSUE TITLE
Weird letter spacing in Integrated Terminal for VS Code

### ISSUE THREAD

**Issue by M16Ali** [microsoft/vscode#120004]
Issue Type: <b>Bug</b>

Hi,

The letter spacing for the integrated terminal is all messed up after an update. I have tried rolling back to an older version as well as resetting any font settings but nothing has worked so far to fix it.

![image](https://user-images.githubusercontent.com/39730473/112713844-6eca7c80-8e94-11eb-98f6-53a6fe975aea.png)

VS Code version: Code - Insiders 1.55.0-insider (d06d2f1d6245ce00b1c36a9cd81a9087d225173e, 2021-03-26T13:59:15.609Z)
OS version: Linux x64 5.8.0-48-generic

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz (8 x 3400)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: disabled_software<br>skia_renderer: enabled_on<br>video_decode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|1, 1, 1|
|Memory (System)|7.65GB (0.54GB free)|
|Process Argv|--no-sandbox --unity-launch --crash-reporter-id 65c1a563-71c6-42b2-b77e-6647a0b3077b|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|ubuntu|
|XDG_CURRENT_DESKTOP|Unity|
|XDG_SESSION_DESKTOP|ubuntu|
|XDG_SESSION_TYPE|x11|
</details><details><summary>Extensions (37)</summary>

Extension|Author (truncated)|Version
---|---|---
numbered-bookmarks|ale|8.0.3
path-intellisense|chr|2.3.0
bracket-pair-colorizer-2|Coe|0.2.0
doxdocgen|csc|1.1.0
vscode-eslint|dba|2.1.19
es7-react-js-snippets|dsz|3.1.1
gitlens|eam|11.3.0
vscode-html-css|ecm|1.10.2
vsc-material-theme|Equ|33.2.0
vsc-material-theme-icons|equ|1.2.2
vscode-pull-request-github|Git|0.24.0
rest-client|hum|0.24.4
better-cpp-syntax|jef|1.15.5
Angular2|joh|11.0.0
vscode-github|Kni|0.30.5
Angular-BeastCode|Mik|10.0.3
csharp|ms-|1.23.9
python|ms-|2021.3.680753044
jupyter|ms-|2021.5.689556049
remote-containers|ms-|0.165.1
remote-ssh|ms-|0.65.1
remote-ssh-edit|ms-|0.65.1
remote-wsl|ms-|0.54.6
vscode-remote-extensionpack|ms-|0.20.0
cpptools|ms-|1.3.0-insiders2
cpptools-extension-pack|ms-|1.0.0
vsliveshare|ms-|1.0.4018
vsliveshare-audio|ms-|0.1.91
vsonline|ms-|1.0.3076
color-highlight|nau|2.3.0
material-icon-theme|PKi|4.6.0
vscode-css-peek|pra|4.1.2
live-sass|rit|3.0.0
LiveServer|rit|5.6.1
code-settings-sync|Sha|3.4.3
vscode-arduino|vsc|0.4.0
JavaScriptSnippets|xab|1.8.0

(2 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstry244:30244315
pythonvsdeb440:30224570
pythonvsded773:30223139
pythonvspyt875:30259475
pythonvsnew554:30265444
pythontb:30258533
pythonvspyt943:30277000
wslfolderdoc:30278405
vspre833:30267464
pythonptprofiler:30281269
vscnewfiletext:30280636
vshan820:30276952
vscoregs2cf2:30281057

```

</details>

**Comment by meganrogge** [comment]
Does resizing the terminal help?

**Comment by M16Ali** [comment]
Yes, I did not notice that before.

**Comment by meganrogge** [comment]
Thanks, working on a fix this iteration

**Comment by Tyriar** [comment]
What we know so far:

- It seems to be related to the WebGL renderer
- It only happens on launch (I think)
- Both cell spacing and height are wrong
- The cursor position seems correct (so the cursor layer has the right dimensions)

**Comment by VanderSP** [comment]
@meganrogge what i dunno if it´s related seems like 3 random bugs of similar nature

**Comment by AwesomeCronk** [comment]
If it is of any use there are a handful of screenshots showcasing this on issue #120483

**Comment by PeterWone** [comment]
> What we know so far:
> 
> * It seems to be related to the WebGL renderer
> * It only happens on launch (I think)
> * Both cell spacing and height are wrong
> * The cursor position seems correct (so the cursor layer has the right dimensions)

I can add

* It only happens when the window is maximised
* A resize fixes it
* Changing the spacing for characters fixes it
* After I rebooted Windows it stopped happening altogether

**Comment by AwesomeCronk** [comment]
For me, it persists between shutdown and startup. Occurs every time vscode launches. 

Could this potentially be worked around by making an extension to change character spacing immediately after launch and then putting it back? It would have to happen before the terminal program (powershell, cmd, etc.) finished starting up to avoid messing with things like winfetch.

**Comment by trig79** [comment]
if it helps anyone, this problem exists for me in the following scenario:
Use split screen in the terminal window i.e. 2 terminals visible.  
Restart VsCode and it boots with one terminal visible only and with crazy letter spacing.
to fix i just click split screen again and all is fine.

**Comment by AwesomeCronk** [comment]
When I get the popup "Extensions have been modified on disk, please reload the window" and I click the reload button, the integrated terminal does this. Clicking `Terminal>New Terminal` after resizing it and then opening a new folder in VSCode starts the terminal without issue.

**Comment by umuluto** [comment]
I also have this issue,  when I set the font to:
```
"terminal.integrated.fontFamily": "monospace"

**Comment by sometimescasey** [comment]
I honestly didn't expect this Googling this to actually help me find the solution to the problem but lo and behold, **resizing the terminal worked**

Thanks everyone! 👀

![image](https://user-images.githubusercontent.com/16248601/116642280-7621f180-a93c-11eb-8680-266515e897a8.png)

**Comment by AwesomeCronk** [comment]
If we rolled back to an older version of WebGL would that potentially fix this?

**Comment by fwaris** [comment]
similar issue. new terminal on 4k monitor has weird spacing. Moving the screen to laptop (1K) fixes it. Move it back to 4K monitor - still ok.

As far as I can tell creating a new terminal on a 4k monitor is causing this for me

**Comment by AwesomeCronk** [comment]
Now *that* sounds like a webGL bug to me... For myself, every time I launch VS Code, it is on a dedicated monitor. I’m going to see if I can try opening it on a secondary monitor run by my integrated gpu instead of my dedicated gpu and see what is happening then.

**Comment by AwesomeCronk** [comment]
Nope, dragging across monitors does nothing for me. I think this is a bug either with WebGL or with the way WebGL is handled in VS Code that manifests on different hardware differently.

**Comment by Tyriar** [comment]
To clarify the monitor workaround, this would only work if your monitors have different window scaling/DPI set. This would fix it because it clears the texture cache since the font size changes. You can also do ctrl/cmd++ or ctrl/cmd+- to force the clearing of the cache.

**Comment by AwesomeCronk** [comment]
Ok, mine are only different resolutions, not different scale. I figured moving the window to a monitor run by a different gpu would do the same, but no dice.

**Comment by Tyriar** [comment]
I believe this will get fixed in https://github.com/microsoft/vscode/pull/123502, for those that experience it often let me know if you hit it again when that is released in insiders tomorrow.

**Comment by Tyriar** [comment]
Just hit this consistently in insiders on my mac, some logs that might be useful if the fix didn't fix it:

<img width="1129" alt="Screen Shot 2021-05-12 at 6 27 47 AM" src="https://user-images.githubusercontent.com/2193314/117982744-2c95b700-b2eb-11eb-9a8b-b09bb2876215.png">

<img width="443" alt="Screen Shot 2021-05-12 at 6 26 54 AM" src="https://user-images.githubusercontent.com/2193314/117982714-24d61280-b2eb-11eb-9d05-19efb2de4f7b.png">
<img width="510" alt="Screen Shot 2021-05-12 at 6 27 04 AM" src="https://user-images.githubusercontent.com/2193314/117982721-269fd600-b2eb-11eb-9470-ddd7fa00c616.png">

xterm is sized correctly at 128x15, selection is fine, the text renderer is stuck at 80x30 though:

<img width="942" alt="Screen Shot 2021-05-12 at 6 28 37 AM" src="https://user-images.githubusercontent.com/2193314/117982871-4a631c00-b2eb-11eb-9cd6-3df855e4d40b.png">

Reducing of the terminal by opening devtools prevents it from happening again, until devtools is hidden. So maybe the width of the terminal is part of the repro?

<img width="920" alt="Screen Shot 2021-05-12 at 6 29 53 AM" src="https://user-images.githubusercontent.com/2193314/117983078-77afca00-b2eb-11eb-8ff0-dded2240f693.png">

I still think this will get fixed by the commit after checking this stuff out.

**Comment by aaajug** [comment]
This is an error caused by the font being used in the VSCode integrated terminal. Other answers say to change it to some default font but I think it's better to just use the same font that your system terminal uses. To fix:

1. Search for the name of the font being used in you system's terminal (for macOS, it's **Menlo**)
2. Open command palette (View -> Command palette)
3. type **settings.json** and click the one that says **Preferences: Open Settings (JSON)**
4. If you haven't modified any preferences in your VSCode before, you should see an almost blank file with just braces {}.
5. Just add this line: `"terminal.integrated.fontFamily": "<SYSTEM TERMINAL FONT>"`

ex. `"terminal.integrated.fontFamily": "Menlo"` (for macOS)

Duplicate of #35681 .

**Comment by Tyriar** [comment]
@aaajug this is actually a different issue since the screenshots show that the font is monospace

**Comment by AmyCox** [comment]
If I extend the terminal window up (make the vertical length of the terminal window longer) it seems to revert the spacing back to normal

**Comment by AkshayRathod07** [comment]
> if it helps anyone, this problem exists for me in the following scenario:
> Use split screen in the terminal window i.e. 2 terminals visible.
> Restart VsCode and it boots with one terminal visible only and with crazy letter spacing.
> to fix i just click split screen again and all is fine.

thanks a lot . its working try this :- 1 - Use split screen in the terminal window i.e. 2 terminals visible.

**Comment by Tyriar** [comment]
@M16Ali the issue is on the May milestone, which means the month in which the work took place. This is coming in 1.57 which will be released in 1-2 days. The month version name is also visible on the website https://code.visualstudio.com/updates/v1_56
